Ai attack/repair basic fix

dnqbob opened the pull request:
@reaperrr 
@abcdefg30

reaperrr commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs:
```diff
@@ -199,12 +177,14 @@ public void Tick(Squad owner)
 
-				if (!ReloadsAutomatically(a))
+				var ammoPools = a.TraitsImplementing<AmmoPool>().ToArray();
+
```
This empty line can be removed.

dnqbob replied:
checkmark

reaperrr commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs:
```diff
@@ -120,5 +120,24 @@ protected static bool IsRearming(Actor a)
+
+			return ammoPools.All(ap => !rearmable.Info.AmmoPools.Contains(ap.Info.Name));
+		}
+
```
This empty line must be removed.

dnqbob replied:
checkmark

dnqbob replied:
I guess I remove it here?

reaperrr commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs:
```diff
@@ -230,7 +210,8 @@ public void Tick(Squad owner)
 			foreach (var a in owner.Units)
 			{
-				if (!ReloadsAutomatically(a) && !FullAmmo(a))
+				var ammoPool = a.TraitsImplementing<AmmoPool>().ToArray();
```
ammoPool**s**.

Also, since we're dropping the `.All` in the methods the `.ToArray()` can go away.

dnqbob replied:
checkmark

dnqbob commented:
So after you check and merge this, I will open several other prs based on bleed.

reaperrr commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs:
```diff
@@ -120,5 +120,24 @@ protected static bool IsRearming(Actor a)
+
+		protected static bool FullAmmo(IEnumerable<AmmoPool> ammoPools)
+		{
+			return ammoPools.All(x => x.HasFullAmmo);
```
This means we always return `false` if any pool doesn't have full ammo, so it would be more performant to do
```
foreach (var ap in ammoPools)
	if (!ap.HasFullAmmo)
		return false;

return true;
```

dnqbob replied:
checkmark

reaperrr commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs:
```diff
@@ -120,5 +120,24 @@ protected static bool IsRearming(Actor a)
+
+		protected static bool HasAmmo(IEnumerable<AmmoPool> ammoPools)
+		{
+			return ammoPools.All(x => x.HasAmmo);
```
Same as above.

dnqbob replied:
checkmark

reaperrr commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs:
```diff
@@ -120,5 +120,24 @@ protected static bool IsRearming(Actor a)
+			if (rearmable == null)
+				return true;
+
+			return ammoPools.All(ap => !rearmable.Info.AmmoPools.Contains(ap.Info.Name));
```
Same here:
```
foreach (var ap in ammoPools)
	if (!rearmable.Info.AmmoPools.Contains(ap.Info.Name))
		return false;

return true;
```

dnqbob replied:
checkmark

reaperrr commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs:
```diff
@@ -199,12 +177,14 @@ public void Tick(Squad owner)
 					continue;
 
-				if (!ReloadsAutomatically(a))
+				var ammoPools = a.TraitsImplementing<AmmoPool>().ToArray();
```
Since we're dropping the `.All` in the methods, the `.ToArray()` can go away.

deleted-user-1 commented:
Please rework the commit messages. I should know from reading the commit messages what type of changes you did and for what reason.

`basic fix on AI state related` can mean anything about AI state. If the types of changes in this commit (visibility/rearming) have nothing in common and that's the reason for the unspecific commit message, consider splitting them into individual commits.

`AirState basic fix` also tells me nothing useful. 

`"IsRearm" fix and name change`

What type of fix?

dnqbob commented:
commit message fixed

deleted-user-1 commented:
I still don't understand why the `"FindEnemy" functions ignore hidden actors` commit has to care about rearming state or what the "IsRearm" fix is about. Not an issue though, thanks for applying the changes.

dnqbob commented:
> I still don't understand why the "FindEnemy" functions ignore hidden actors commit has to care about rearming state or what the "IsRearm" fix is about. Not an issue though, thanks for applying the changes.

ummm... because title is short so I have to discribe it within words, probably crude. Further detail can be seen in files so I think it is unecessary here. 

Thanks your advice anyway.

reaperrr commented on OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs:
```diff
@@ -160,12 +175,14 @@ void IBotTick.BotTick(IBot bot)
 		{
-			return World.FindActorsInCircle(pos, radius).Where(IsEnemyUnit).ClosestTo(pos);
+			var units = World.FindActorsInCircle(pos, radius).Where(IsEnemyUnit);
+			return units.Where(IsNotHiddenUnit).ClosestTo(pos);
```
This is only a question for now, not a change request (yet): Is there a reason why this doesn't do the same ` ?? units.ClosestTo(pos)` as the overload above?

dnqbob replied:
Because find actors in `FindClosestEnemy(WPos pos, WDist radius)` is often used for target scanning within an area. If we apply `?? units.ClosestTo(pos)` here then AI will try to attack it, when all other not-hidden units are killed (but stucked because it cannot be attacked when cloaked). 

`FindClosestEnemy(WPos pos)` is often used for approaching target for AI squad. So `?? units.ClosestTo(pos)` here means if AI really cannot find any enemy **on map** that not cloaked, they will finally try to approach the cloacked one.

reaperrr replied:
In that case it should be possible to combine the two `Where()` though, since both need to be checked anyway. Then we also wouldn't need the `var` and could just return
```
World.FindActorsInCircle(pos, radius).Where(a => IsEnemyUnit(a) && IsNotHiddenUnit(a)).ClosestTo(pos)
```
which should be a bit more performance-efficient.

dnqbob replied:
Fixed

reaperrr commented:
> I still don't understand why the "FindEnemy" functions ignore hidden actors commit has to care about rearming state

I agree It would be better if the `IsRearm` moving to `StateBase` from commit 1 was merged into commit 2, yes.

>  or what the "IsRearm" fix is about

I guess it's about the type lookup no longer being pulled out into a var (since that was redundant), but I agree it's so minor that it would be less confusing if the 'fix' part was removed from the commit message.

dnqbob commented:
Ah, my mistakes here. I use this branch as base branch yesterday but never check the commit problem.

dnqbob committed:
StateBase.cs modified:

1. Optimize & move "ammo" related function from "AirStates.cs" to StateBase.cs

2. Optimize & move "IsRearm" function from "AirStates.cs" to StateBase.cs, name changed to "IsRearming"

(optimized by reaperrr)

dnqbob commented:
.... please give me sometime here, It seems I mistake something again for another place.

dnqbob commented:
Need to learn git more for me.

reaperrr reviewed (dismissed):
Code looks good to me now, didn't spot any issues in a 4-AI RA game, either.

dnqbob commented:
dude,  you can also try lay some mines in front of AI's door,  and see if they will stuck when attack.  

There are still some lines needs to be fixed in states because they directly use `isEnemyUnit` to detect enemy,  **if we want ignore hidden enemy works normally, but not for now**. Unfixed lines will let AttackMove state switch to Attack state when encounter the hidden unit, although they will begin to search and attack visible one during Attack state.

abcdefg30 commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs:
```diff
@@ -220,12 +177,13 @@ public void Tick(Squad owner)
 
-				if (!ReloadsAutomatically(a))
+				var ammoPools = a.TraitsImplementing<AmmoPool>();
+				if (!ReloadsAutomatically(ammoPools, a.TraitOrDefault<Rearmable>()))
```
I must have missed this during all the back and forth. Can someone elaborate why we're suddenly doing the trait lookups outside of the method (= high coupling)? Since we're not `ToArray`ing or `ToList`ing we'll still enumerate twice worst case, so there should basically be no difference between doing the lookups here and doing it inside the method (except coupling).

dnqbob replied:
ummmm.. yeah I also said readability is down here in exchange of the performance.

dnqbob replied:
But it will require more thinking on probably change all those functions back for readability and low coupling, since they are all about the ammo related.

abcdefg30 replied:
My point was that this doesn't look like it gives us any performance gains.

reaperrr replied:
The idea was that both `ReloadsAutomatically` and `HasAmmo()` below were looking up `a.TraitsImplementing<AmmoPool>();` before, so I was assuming that caching it once, passing it through as argument and then running a foreach on them would save at least one of the (up to) two look-ups as well as the additional enumerations.

I'm not that firm on the topic though, if you say passing them as array would be more efficient we should do that, of course.

dnqbob replied:
Changed to `ToArray`

abcdefg30 commented on OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs:
```diff
@@ -160,12 +175,14 @@ void IBotTick.BotTick(IBot bot)
 		{
-			return World.FindActorsInCircle(pos, radius).Where(IsEnemyUnit).ClosestTo(pos);
+			var units = World.FindActorsInCircle(pos, radius).Where(IsEnemyUnit);
+			return units.Where(IsNotHiddenUnit).ClosestTo(pos);
```
Why was this change reverted again?

dnqbob replied:
Oh...ummm for fixing the first commit I drop the second then copy a back up one

dnqbob replied:
Should be fixed

abcdefg30 commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs:
```diff
@@ -251,9 +209,10 @@ public void Tick(Squad owner)
 			foreach (var a in owner.Units)
 			{
-				if (!ReloadsAutomatically(a) && !FullAmmo(a))
+				var ammoPools = a.TraitsImplementing<AmmoPool>();
```
This has the same problem.

Fwiw, I'm still not a fan of the approach, but if you both think it is worth it we can keep it.

dnqbob replied:
Ummm, I quite don't know how many performance it will bring. But AI tick can sometimes lag the game (mainly pathfinding, though). 

I just finished a GroundAttackMove state fix up to #17327(thanks tovl's idea), in my test map (a lot of twisted road) the game will lag everytime when AI attackmove. AI airforce can also lags for some unknown reasons when there are many aircrafts (they don't even do complex pathfinding). I don't know will improvement of some places can help on the lags caused by AI.

dnqbob replied:
should be fixed

dnqbob committed:
"FindEnemy" functions ignore hidden actors

(helped by reaperrr and abcdefg30)

abcdefg30 commented on OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs:
```diff
@@ -126,6 +126,21 @@ public bool IsEnemyUnit(Actor a)
+		public bool IsNotHiddenUnit(Actor a)
+		{
+			var hasModifier = false;
+			var visModifiers = a.TraitsImplementing<IVisibilityModifier>().ToArray();
```
Now this `ToArray` is one too many again, sorry.

dnqbob replied:
It is OK here. I also update my back up.

abcdefg30 reviewed (commented):
Some explanation to when we should use `ToArray`:
`TraitsImplementing` returns an `IEnumerable`. That means it does basically nothing until you start looping over the enumerable. While looping the next element is determined/fetched for each loop run individually. When we loop a second time over the entire enumerable this happens again.
Now, calling `ToArray` or `ToList` forces the enumerable to do this right away; to determine all elements and save them somewhere to memory. Iterating over that list/array will now be much faster than fetching each element again like before.
Tl;dr: Using `IEnumerable` == do the work when we actually run through a loop or similar, while `ToList`ing/`ToArray`ing == do the work right away but be faster when we loop later.

If we only ever loop once directly after creating the enumerable you should usually not use `ToArray` or `ToList` since that has the overhead of allocating memory (and in theory a second iteration: The first is the iteration when using ToArray, and the second for looping over that array).
If you loop several times you should use `ToArray` or `ToList` since the array/list enumerators are much faster than evaluating the `IEnumerable` again each time.

abcdefg30 commented:
Some links if you are interested:
https://stackoverflow.com/questions/23536541/performance-between-iterating-through-ienumerablet-and-listt
https://stackoverflow.com/questions/3628425/ienumerable-vs-list-what-to-use-how-do-they-work

dnqbob commented:
> Some links if you are interested:
> https://stackoverflow.com/questions/23536541/performance-between-iterating-through-ienumerablet-and-listt
> https://stackoverflow.com/questions/3628425/ienumerable-vs-list-what-to-use-how-do-they-work

Marked.

dnqbob commented:
Does this mean, the `IEnumerable` is just a placeholder which later may be used (in Linq), every time it is being used will result in an extra process of converting to list/array?

```
// IEn is a Linq based "IEnumerable"
X1 = IEn.Select(...).Where(...);
X2 = X1.Select(...).Where(...);

Y1 = IEn.Select(...).Where(...).ToArray();
```
 X1 and X2 are nothing but placeholders, Y is actually an array?

abcdefg30 commented:
> X1 and X2 are nothing but placeholders, Y is actually an array?

You can put it like that, yes.

> Does this mean, the `IEnumerable` is just a placeholder which later may be used (in Linq), every time it is being used will result in an extra process of converting to list/array?

"Placeholder which we use later" yes. But it is not converted to a list/an array. I think you got the idea though: Every time it is used we will need to fetch all the elements again. (This just doesn't happen all at once, but after each time we ran through a `foreach` body the next element is fetched. So you basically do "fetch next element - run loop body - fetch next element - run loop body - ..." afaik.)

abcdefg30 reviewed (approved):
Ok, lgtm.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/92d5fa562493b758f23cdc97bdc22e04c8d4459f...114f2b372a2f419b4943d8de354f19a7cf17d5a1)
